[SPARK-21867][CORE] Support async spilling in UnsafeShuffleWriter

ericvandenbergfb-zz opened the pull request:
## What changes were proposed in this pull request?

Add a multi-shuffle sorter which supports asynchronous spilling during a shuffle external sort.  The benefit is that we can insert and sort/spill in parallel, reducing the overall latency for jobs that are heavy on shuffling. The multi-shuffle sorter is added between the UnsafeShuffleWriter and ShuffleExternalSorter such that few changes are needed outside of this component, and as such, we can see clearly there is little room for regressing other code paths.

The multi-shuffle sorter is enabled via a configuration flag, spark.shuffle.async.num.sorter (default 1)  If the value is 1, then the multi-shuffle sorter is not used, it must be configured to have multiple
sorters (>=2)

There is a design spec here attached to the jira.

## How was this patch tested?

Added unit tests specifically for the MultiShuffleSorter to exercise under various spill and insert conditions.
Extended the UnsafeShuffleWriterSuite to run against a single ShuffleExternalSorter or multiple via the MultiShuffleSorter to ensure no regressions.
Ran against production work loads and observed gains and validated based on logging.
One of the larger jobs went from ~6030 cpu-hr to ~4298 cpu-hr which was a total aggregate reduction of ~29%.  For larger jobs, this provides a great tuning option to reduce job time.

Commit:
[SPARK-21867][CORE] Support async spilling in UnsafeShuffleWriter

Add a multi-shuffle sorter which supports asynchronous spilling during a shuffle external sort.
The benefit is that we can insert and sort/spill in parallel, reducing the overall latency for jobs that
are heavy on shuffling (as are many ads jobs). The multi-shuffle sorter is added between the UnsafeShuffleWriter
and ShuffleExternalSorter such that few changes are needed outside of this component, and as such, we can
see clearly there is little room for regressing other code paths.

The multi-shuffle sorter is enabled via a configuration flag, spark.shuffle.async.num.sorter (default 1)
If the value is 1, then the multi-shuffle sorter is not used, it must be configured to have multiple
sorters (>=2)

There is a design spec here attached to the jira.

AmplabJenkins commented:
Can one of the admins verify this patch?

vanzin commented:
ok to test

SparkQA commented:
**[Test build #84800 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84800/testReport)** for PR 19955 at commit [`7f751ec`](https://github.com/apache/spark/commit/7f751ec23ba6d8c53c009edb7d62a460e9166d7f).

SparkQA commented:
**[Test build #84800 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84800/testReport)** for PR 19955 at commit [`7f751ec`](https://github.com/apache/spark/commit/7f751ec23ba6d8c53c009edb7d62a460e9166d7f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class MultiShuffleSorter extends ShuffleSorter `
  * `final class ShuffleExternalSorter extends ShuffleSorter `
  * `abstract class ShuffleSorter extends MemoryConsumer `
  * `public class ShuffleSorterFactory `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/84800/
Test FAILed.

Commit:
Update return type in Mockito usage in test.
Update async code to use ReentrantLock and Condition.

SparkQA commented:
**[Test build #84889 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84889/testReport)** for PR 19955 at commit [`a3ea93c`](https://github.com/apache/spark/commit/a3ea93cd829debc13de6aac3fc94e49c3a03dabc).

SparkQA commented:
**[Test build #84889 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84889/testReport)** for PR 19955 at commit [`a3ea93c`](https://github.com/apache/spark/commit/a3ea93cd829debc13de6aac3fc94e49c3a03dabc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/84889/
Test FAILed.

gatorsmile commented:
cc @JoshRosen @jiangxb1987

cloud-fan commented:
shall we do it for all the external sorters?

cloud-fan commented:
and also what's the behavior if the memory is used up but `spark.shuffle.async.num.sorter` is not hit yet?

sitalkedia commented:
@ericvandenbergfb - In the testing section, let's put the benchmark numbers for large shuffle heavy jobs we ran internally.

AmplabJenkins commented:
Can one of the admins verify this patch?

ericvandenbergfb commented:
@cloud-fan I think this could be enabled for other sorters (mentioned that at end of design spec), just wanted to get this in incrementally.

ericvandenbergfb commented:
@cloud-fan In this case it reverts to the existing behavior, it will spill synchronously, free memory and then continue inserting.   In addition, it would re-tune (down) the spill threshold so as it likely not hit this condition next time.

ericvandenbergfb commented:
@sitalkedia Added comment in test section on some benchmarks.

Commit:
Update synchronization lock to avoid race condition.

SparkQA commented:
**[Test build #84971 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84971/testReport)** for PR 19955 at commit [`171f1bb`](https://github.com/apache/spark/commit/171f1bbe4f12171ab6c0d2e30da184bd58ede94f).

SparkQA commented:
**[Test build #84971 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84971/testReport)** for PR 19955 at commit [`171f1bb`](https://github.com/apache/spark/commit/171f1bbe4f12171ab6c0d2e30da184bd58ede94f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/84971/
Test FAILed.

Commit:
Change visibility of mocking classes and sub-class (rather than abstract
base class) of mocked class.

Note the failures aren't repro'ing local on my machine, so I'm going
based off the failure stack traces.

SparkQA commented:
**[Test build #84983 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84983/testReport)** for PR 19955 at commit [`59e7720`](https://github.com/apache/spark/commit/59e7720cf0895d4359decdee57eec6fc11bc2fe0).

SparkQA commented:
**[Test build #84983 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84983/testReport)** for PR 19955 at commit [`59e7720`](https://github.com/apache/spark/commit/59e7720cf0895d4359decdee57eec6fc11bc2fe0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class MultiShuffleSorter extends ShuffleSorter `
  * `public class ShuffleExternalSorter extends ShuffleSorter `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/84983/
Test FAILed.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
ok to test

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
ping @ericvandenbergfb to update or close.

AmplabJenkins commented:
Can one of the admins verify this patch?
